File: github-tree.py
# github-tree.py
# by coldnw.t.me
# github.com/rhythmcache

import requests
import sys
import os
from urllib.parse import urlparse

# Add your GitHub API token here if available
API_TOKEN = ""  # Replace with "TOKEN-VALUE" if needed

def fetch_repo_info(user, repo):
    url = f"https://api.github.com/repos/{user}/{repo}"
    headers = get_headers()
    response = requests.get(url, headers=headers)
    if response.status_code == 200:
        return response.json()
    else:
        print("Error fetching repository info.")
        return None

def fetch_tree_data(user, repo, branch):
    url = f"https://api.github.com/repos/{user}/{repo}/git/trees/{branch}?recursive=1"
    headers = get_headers()
    all_items = []

    while url:
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            data = response.json()
            all_items.extend(data["tree"])
            url = response.links.get('next', {}).get('url')  # Pagination
        elif response.status_code == 403:
            print("Rate limit exceeded. Try again later.")
            return None
        else:
            print(f"Error fetching tree: {response.status_code}")
            print(response.json().get("message", "No additional information provided."))
            return None

    return all_items

def parse_github_url(repo_url):
    if repo_url.endswith(".git"):
        repo_url = repo_url[:-4]
    
    parsed_url = urlparse(repo_url)
    if not parsed_url.netloc.endswith("github.com"):
        return None, None

    path_parts = parsed_url.path.strip("/").split("/")
    if len(path_parts) < 2:
        return None, None

    return path_parts[:2]

def create_placeholder_structure(tree, base_path):
    for item in tree:
        path = os.path.join(base_path, item["path"])
        if item["type"] == "tree":
            os.makedirs(path, exist_ok=True)
        elif item["type"] == "blob":
            os.makedirs(os.path.dirname(path), exist_ok=True)
            with open(path, "w") as f:
                pass

    print(f"Placeholder structure created at: {base_path}")

def print_tree(tree):
    structure = {}

    for item in tree:
        parts = item["path"].split("/")
        current_level = structure
        for part in parts:
            current_level = current_level.setdefault(part, {})

    def print_nested(d, prefix=""):
        for i, (key, value) in enumerate(d.items()):
            connector = "└── " if i == len(d) - 1 else "├── "
            print(prefix + connector + key)
            print_nested(value, prefix + ("    " if connector == "└── " else "│   "))

    print_nested(structure)

def count_files_and_folders(tree):
    files = sum(1 for item in tree if item["type"] == "blob")
    folders = sum(1 for item in tree if item["type"] == "tree")
    return files, folders

def get_headers():
    """Prepare headers for the request, including API token if available."""
    headers = {}
    if API_TOKEN:
        print("Using API token to fetch data...")
        headers["Authorization"] = f"token {API_TOKEN}"
    return headers

def main():
    if len(sys.argv) < 3:
        print("Usage:")
        print(f"  python {os.path.basename(__file__)} touch <output-path> <GitHub Repository URL> [branch]")
        print(f"  python {os.path.basename(__file__)} view <GitHub Repository URL> [branch]")
        return

    command = sys.argv[1]
    if command not in ["touch", "view"]:
        print(f"Error: Unsupported command '{command}'.")
        return

    if command == "touch":
        if len(sys.argv) < 4:
            print("Error: Missing arguments for 'touch' command.")
            return
        
        base_path = sys.argv[2]
        repo_url = sys.argv[3]
        branch = sys.argv[4] if len(sys.argv) > 4 else None
        user, repo = parse_github_url(repo_url)

        if not user or not repo:
            print("Error: Invalid GitHub URL.")
            return

        if not branch:
            repo_info = fetch_repo_info(user, repo)
            branch = repo_info.get("default_branch", "main") if repo_info else "main"

        tree = fetch_tree_data(user, repo, branch)
        if tree:
            create_placeholder_structure(tree, base_path)

    elif command == "view":
        if len(sys.argv) < 3:
            print("Error: Missing arguments for 'view' command.")
            return
        
        repo_url = sys.argv[2]
        branch = sys.argv[3] if len(sys.argv) > 3 else None
        user, repo = parse_github_url(repo_url)

        if not user or not repo:
            print("Error: Invalid GitHub URL.")
            return

        if not branch:
            repo_info = fetch_repo_info(user, repo)
            branch = repo_info.get("default_branch", "main") if repo_info else "main"

        tree = fetch_tree_data(user, repo, branch)
        if tree:
            print_tree(tree)
            files, folders = count_files_and_folders(tree)
            print(f"\nTotal folders: {folders}")
            print(f"Total files: {files}")

if __name__ == "__main__":
    main()
    

File: github-tree.go
package main

import (
    "encoding/json"
    "flag"
    "fmt"
    "io/ioutil"
    "net/http"
    "os"
    "path/filepath"
    "strings"
)

type TreeResponse struct {
    Tree []TreeItem `json:"tree"`
}

type TreeItem struct {
    Path string `json:"path"`
    Type string `json:"type"`
}

type RepoInfo struct {
    DefaultBranch string `json:"default_branch"`
}

func parseGitHubURL(repoURL string) (string, string) {
    repoURL = strings.TrimSuffix(repoURL, ".git")
    parts := strings.Split(repoURL, "/")
    if len(parts) < 2 {
        return "", ""
    }
    return parts[len(parts)-2], parts[len(parts)-1]
}

func fetchRepoInfo(user, repo, pat string) (*RepoInfo, error) {
    url := fmt.Sprintf("https://api.github.com/repos/%s/%s", user, repo)
    client := &http.Client{}
    req, err := http.NewRequest("GET", url, nil)
    if err != nil {
        return nil, err
    }

    if pat != "" {
        req.Header.Add("Authorization", "token "+pat)
    }
    
    resp, err := client.Do(req)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    if resp.StatusCode != 200 {
        return nil, fmt.Errorf("API error: %d", resp.StatusCode)
    }

    var info RepoInfo
    if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
        return nil, err
    }
    return &info, nil
}

func fetchTreeData(user, repo, branch, pat string) ([]TreeItem, error) {
    url := fmt.Sprintf("https://api.github.com/repos/%s/%s/git/trees/%s?recursive=1", user, repo, branch)
    client := &http.Client{}
    req, err := http.NewRequest("GET", url, nil)
    if err != nil {
        return nil, err
    }

    if pat != "" {
        req.Header.Add("Authorization", "token "+pat)
    }

    resp, err := client.Do(req)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    if resp.StatusCode != 200 {
        return nil, fmt.Errorf("API error: %d", resp.StatusCode)
    }

    var treeResp TreeResponse
    if err := json.NewDecoder(resp.Body).Decode(&treeResp); err != nil {
        return nil, err
    }
    return treeResp.Tree, nil
}

func createPlaceholderStructure(tree []TreeItem, basePath string) error {
    for _, item := range tree {
        path := filepath.Join(basePath, item.Path)
        if item.Type == "tree" {
            if err := os.MkdirAll(path, 0755); err != nil {
                return err
            }
        } else if item.Type == "blob" {
            dir := filepath.Dir(path)
            if err := os.MkdirAll(dir, 0755); err != nil {
                return err
            }
            if err := ioutil.WriteFile(path, []byte{}, 0644); err != nil {
                return err
            }
        }
    }
    return nil
}

type Node struct {
    Name     string
    Children map[string]*Node
}

func printTree(tree []TreeItem) {
    root := &Node{Children: make(map[string]*Node)}
    
    // Build tree structure
    for _, item := range tree {
        parts := strings.Split(item.Path, "/")
        current := root
        for _, part := range parts {
            if current.Children[part] == nil {
                current.Children[part] = &Node{
                    Name:     part,
                    Children: make(map[string]*Node),
                }
            }
            current = current.Children[part]
        }
    }

    // Print tree structure
    var printNode func(*Node, string, bool)
    printNode = func(node *Node, prefix string, isLast bool) {
        if node.Name != "" {
            connector := "├── "
            if isLast {
                connector = "└── "
            }
            fmt.Println(prefix + connector + node.Name)
        }

        keys := make([]string, 0, len(node.Children))
        for k := range node.Children {
            keys = append(keys, k)
        }

        for i, key := range keys {
            newPrefix := prefix
            if node.Name != "" {
                if isLast {
                    newPrefix += "    "
                } else {
                    newPrefix += "│   "
                }
            }
            printNode(node.Children[key], newPrefix, i == len(keys)-1)
        }
    }

    printNode(root, "", true)
}

func countFilesAndFolders(tree []TreeItem) (int, int) {
    files, folders := 0, 0
    for _, item := range tree {
        if item.Type == "blob" {
            files++
        } else if item.Type == "tree" {
            folders++
        }
    }
    return files, folders
}

func main() {
    touchCmd := flag.NewFlagSet("touch", flag.ExitOnError)
    touchPath := touchCmd.String("path", "", "Output path for the placeholder structure")
    touchPat := touchCmd.String("pat", "", "GitHub Personal Access Token")

    viewCmd := flag.NewFlagSet("view", flag.ExitOnError)
    viewPat := viewCmd.String("pat", "", "GitHub Personal Access Token")

    if len(os.Args) < 2 {
        fmt.Println("Expected 'touch' or 'view' subcommands")
        os.Exit(1)
    }

    switch os.Args[1] {
    case "touch":
        touchCmd.Parse(os.Args[2:])
        args := touchCmd.Args()
        if len(args) < 1 || *touchPath == "" {
            fmt.Println("Usage: github-tree touch -path=<output-path> -pat=<token> <repo-url> [branch]")
            os.Exit(1)
        }

        repoURL := args[0]
        branch := ""
        if len(args) > 1 {
            branch = args[1]
        }

        user, repo := parseGitHubURL(repoURL)
        if user == "" || repo == "" {
            fmt.Println("Invalid GitHub URL")
            os.Exit(1)
        }

        if branch == "" {
            info, err := fetchRepoInfo(user, repo, *touchPat)
            if err != nil {
                fmt.Printf("Error fetching repo info: %v\n", err)
                os.Exit(1)
            }
            branch = info.DefaultBranch
        }

        tree, err := fetchTreeData(user, repo, branch, *touchPat)
        if err != nil {
            fmt.Printf("Error fetching tree data: %v\n", err)
            os.Exit(1)
        }

        if err := createPlaceholderStructure(tree, *touchPath); err != nil {
            fmt.Printf("Error creating structure: %v\n", err)
            os.Exit(1)
        }
        fmt.Printf("Placeholder structure created at: %s\n", *touchPath)

    case "view":
        viewCmd.Parse(os.Args[2:])
        args := viewCmd.Args()
        if len(args) < 1 {
            fmt.Println("Usage: github-tree view -pat=<token> <repo-url> [branch]")
            os.Exit(1)
        }

        repoURL := args[0]
        branch := ""
        if len(args) > 1 {
            branch = args[1]
        }

        user, repo := parseGitHubURL(repoURL)
        if user == "" || repo == "" {
            fmt.Println("Invalid GitHub URL")
            os.Exit(1)
        }

        if branch == "" {
            info, err := fetchRepoInfo(user, repo, *viewPat)
            if err != nil {
                fmt.Printf("Error fetching repo info: %v\n", err)
                os.Exit(1)
            }
            branch = info.DefaultBranch
        }

        tree, err := fetchTreeData(user, repo, branch, *viewPat)
        if err != nil {
            fmt.Printf("Error fetching tree data: %v\n", err)
            os.Exit(1)
        }

        printTree(tree)
        files, folders := countFilesAndFolders(tree)
        fmt.Printf("\nTotal folders: %d\n", folders)
        fmt.Printf("Total files: %d\n", files)

    default:
        fmt.Printf("Unknown command: %s\n", os.Args[1])
        os.Exit(1)
    }
}
